### DISCUSSION TITLE
webSocketImpl on Cloudflare Workers

### DISCUSSION THREAD

**Discussion by jeevcat** [enisdenjo/graphql-ws#376] [Q&A]
I'm trying to create a Client for integration testing purposes in a Cloudflare Workers environment.

If I pass the "WebSocket" implementation [provided by Cloudflare Workers](https://github.com/cloudflare/workers-types/blob/f661c288e64198fe7744a7118a61ada1320fd0b8/index.d.ts#L1881-L1894), it fails the `isWebSocket` type guard, due to different naming of the members `CLOSED`, `CLOSING`, etc.

How should I work around this? Create some kind of wrapper class?

Thanks for your input!

**Comment by enisdenjo** [comment]
I don't think you can use Cloudflare Workers WebSocket as a client, [only as a server](https://developers.cloudflare.com/workers/learning/using-websockets/#writing-a-websocket-server).

In case you need a server example, you can check it out here: [cloudflare-worker-graphql-ws-template](https://github.com/enisdenjo/cloudflare-worker-graphql-ws-template).

**Reply by jeevcat** [reply]
Thanks for your response! Further down in your link it actually does mention using Cloudflare Works WebSocket as a client https://developers.cloudflare.com/workers/learning/using-websockets/#writing-a-websocket-client

**Reply by enisdenjo** [reply]
My bad @jeevcat.

Since the Cloudflare Workers WebSocket is different from the web standard implementation, a wrapper is indeed necessary. [Their WebSocket](https://github.com/cloudflare/workers-types/blob/38b7e0fba83f01654a00b0d805cd01211a419f3d/index.d.ts#L1881-L1894) has to be implemented in accordance with [MDN WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket).  

A version adapted exclusively for `graphql-ws` would look like this (**untested**):
```ts
import "@cloudflare/workers-types";

class WebSocketForGraphQLWS {
  private ws: WebSocket;
  constructor(url: string, protocol: string) {
    (async () => {
      const res = await fetch(url, {
        headers: {
          Upgrade: "websocket",
          "Sec-WebSocket-Protocol": protocol,
        },
      });

      this.ws = res.webSocket;
      if (!this.ws) {
        throw new Error("server didn't accept WebSocket");
      }

      this.ws.accept();

      this.ws.addEventListener("open", this.onopen);
      this.ws.addEventListener("error", this.onerror);
      this.ws.addEventListener("close", this.onclose);
      this.ws.addEventListener("message", this.onmessage);
    })().catch((e) => this.onerror(e));
  }

  // should be implemented by graphql-ws in timely manner
  public onopen() {
    throw new Error("onopen is not implemented");
  }
  public onerror() {
    throw new Error("onerror is not implemented");
  }
  public onclose() {
    throw new Error("onclose is not implemented");
  }
  public onmessage() {
    throw new Error("onmessage is not implemented");
  }

  // properties and methods used by graphql-ws
  static CONNECTING = WebSocket.READY_STATE_CONNECTING;
  static OPEN = WebSocket.READY_STATE_OPEN;
  static CLOSING = WebSocket.READY_STATE_CLOSING;
  static CLOSED = WebSocket.READY_STATE_CLOSED;
  get readyState() {
    return this.ws.readyState;
  }
  public send(msg: string) {
    this.ws.send(msg);
  }
  public close(code: number, reason: string) {
    this.ws.close(code, reason);
  }
}
```

Would be lovely if you can test this out and report back, I'd like to add it to the documentation. Thanks in advance!

**Reply by tornqvist** [reply]
I stumbled on the same problem, using a Cloudflare Worker as a gateway for another GraphQL API. Using the `fetch` approach as above didn't work as the worker would throw an error that it couldn't fetch a `ws://` url. However, using the `WebSocket` as is and proxying the static props and event listeners did work!

But there's a caveat! There's a discrepancy between the Cloudflare production environment and local development using [Miniflare](https://github.com/cloudflare/miniflare). Miniflare doesn't support using the `on`-methods to attach listens, but Cloudflare does. Using [vite](https://vitejs.dev) to compile the worker, this is what I came up with which actually does work, both in production and locally. Using [Cloudflare Environments](https://developers.cloudflare.com/workers/platform/environments) would though probably be more idiomatic.

```js
class WebSocketForGraphQLWS extends WebSocket {
  constructor(url, protocols) {
    super(url, protocols)

    if (import.meta.env.DEV) {
      this.addEventListener('open', (...args) => this.onopen?.(...args))
      this.addEventListener('error', (...args) => this.onerror?.(...args))
      this.addEventListener('close', (...args) => this.onclose?.(...args))
      this.addEventListener('message', (...args) => this.onmessage?.(...args))
    }
  }

  static CONNECTING = WebSocket.READY_STATE_CONNECTING
  static OPEN = WebSocket.READY_STATE_OPEN
  static CLOSING = WebSocket.READY_STATE_CLOSING
  static CLOSED = WebSocket.READY_STATE_CLOSED
}
```
